### PR TITLE
🔥 Feature: add ability to print custom message on startup

### PR DESCRIPTION
**Pull request by Saman-Safaei**
add a function in listener.go to print out the developer provided custom message

### CODE REVIEW

**Comment by efectn** [comment]
I think it's better idea to improve onListen hook instead of adding new config

**Comment by gaby** [comment]
> I think it's better idea to improve onListen hook instead of adding new config

Agree

**Comment by cmd777** [comment]
> I think it's better idea to improve onListen hook instead of adding new config

@efectn So, something like this? (this is just a mock-up, and does not take prefork into account)

```go
func (app *App) SetStartupMessage(fun ...func() error) {
	// disable default startup message
	app.config.DisableStartupMessage = true

	// add each func to the onListen hook
	app.hooks.onListen = append(app.hooks.onListen, fun...)
}
```
and it allows for this:

```go
app := fiber.New()

app.SetStartupMessage(
	func() error {
		log.Println("Hello, ")
		return nil
	}, func() error {
		log.Println("World!")
		return nil
	},
)

log.Fatal(app.Listen(":9090"))

2023/06/02 15:03:13 Hello, 
2023/06/02 15:03:13 World!
^Csignal: interrupt
```

**Comment by efectn** [comment]
> > I think it's better idea to improve onListen hook instead of adding new config
> 
> @efectn So, something like this? (this is just a mock-up, and does not take prefork into account)
> 
> ```go
> func (app *App) SetStartupMessage(fun ...func() error) {
> 	// disable default startup message
> 	app.config.DisableStartupMessage = true
> 
> 	// add each func to the onListen hook
> 	app.hooks.onListen = append(app.hooks.onListen, fun...)
> }
> ```
> 
> and it allows for this:
> 
> ```go
> app := fiber.New()
> 
> app.SetStartupMessage(
> 	func() error {
> 		log.Println("Hello, ")
> 		return nil
> 	}, func() error {
> 		log.Println("World!")
> 		return nil
> 	},
> )
> 
> log.Fatal(app.Listen(":9090"))
> 
> 2023/06/02 15:03:13 Hello, 
> 2023/06/02 15:03:13 World!
> ^Csignal: interrupt
> ```



> > I think it's better idea to improve onListen hook instead of adding new config
> 
> @efectn So, something like this? (this is just a mock-up, and does not take prefork into account)
> 
> ```go
> func (app *App) SetStartupMessage(fun ...func() error) {
> 	// disable default startup message
> 	app.config.DisableStartupMessage = true
> 
> 	// add each func to the onListen hook
> 	app.hooks.onListen = append(app.hooks.onListen, fun...)
> }
> ```
> 
> and it allows for this:
> 
> ```go
> app := fiber.New()
> 
> app.SetStartupMessage(
> 	func() error {
> 		log.Println("Hello, ")
> 		return nil
> 	}, func() error {
> 		log.Println("World!")
> 		return nil
> 	},
> )
> 
> log.Fatal(app.Listen(":9090"))
> 
> 2023/06/02 15:03:13 Hello, 
> 2023/06/02 15:03:13 World!
> ^Csignal: interrupt
> ```

Hooks handlers accept parameters as you can see here https://github.com/gofiber/fiber/blob/master/hooks.go#L8

We can pass needed variables to hook like this. However, it'll break handler signature. Perhaps we put them into a struct and pass it as variadic, so it won't break compatibility

**Comment by efectn** [comment]
I'd recommend something like this:

```go
type ListenData struct {
	Host string
	Port string
	TLS  string
}

app.Hooks().OnListen(func(data ...ListenData) error {
	if fiber.IsChild() {
		return nil
	}

	fmt.Print(data[0].Host)

	return nil
})
```

File: app.go [hunk 1]
```diff
@@ -1091,7 +1091,11 @@ func (app *App) serverErrorHandler(fctx *fasthttp.RequestCtx, err error) {
 }
 
 // startupProcess Is the method which executes all the necessary processes just before the start of the server.
-func (app *App) startupProcess() *App {
+func (app *App) startupProcess(listenData ...ListenData) *App {
+	if err := app.hooks.executeOnListenHooks(listenData...); err != nil {
```
Comment (ReneWerner87) [inline app.go:1095]: pls use the new function 
runOnListenHooks for this

**Review by ReneWerner87 — CHANGES_REQUESTED** [review]
check my hint

**Comment by Saman-Safaei** [comment]
> check my hint

I fixed that

File: docs/guide/hooks.md [hunk 2]
```diff
@@ -127,12 +127,21 @@ OnListen is a hook to execute user functions on Listen, ListenTLS, Listener.
...
+<Tabs>
+<TabItem value="onlisten-example" label="OnListen Example">
+
+```go
+app := fiber.New(fiber.Config{
+  DisableStartupMessage: true,
+})
+
 app.Hooks().OnListen(func(listenData fiber.ListenData) error {
   fiber.IsChild()
```
Comment (cmd777) [inline docs/guide/hooks.md:139]: this is missing an !if (assuming you want to print once)

File: docs/guide/hooks.md [hunk 3]
```diff
@@ -127,12 +127,21 @@ OnListen is a hook to execute user functions on Listen, ListenTLS, Listener.
...
   fmt.Println(listenData.TLS + "://" + listenData.Host + ":" + listenData.Port)
```
Comment (cmd777) [inline docs/guide/hooks.md:140]: This is just preference, but should use log.Println instead

**Review by cmd777 — CHANGES_REQUESTED** [review]

File: docs/guide/hooks.md [hunk 4]
```diff
@@ -127,6 +127,29 @@ OnListen is a hook to execute user functions on Listen, ListenTLS, Listener.
...
+<Tabs>
+<TabItem value="onlisten-example" label="OnListen Example">
+
+```go
+app := fiber.New(fiber.Config{
+  DisableStartupMessage: true,
+})
+
+app.Hooks().OnListen(func(listenData fiber.ListenData) error {
+  if (fiber.IsChild()) {
```
Comment (cmd777) [inline docs/guide/hooks.md:139]: Does work, but normally gets shortened to
```suggestion
  if fiber.IsChild() {
```
if you use VSCode or similar

File: listen.go [hunk 5]
```diff
@@ -231,8 +235,26 @@ func (app *App) ListenMutualTLSWithCertificate(addr string, cert tls.Certificate
 	return app.server.Serve(ln)
 }
 
+// prepareListenData create an slice of ListenData
+func (app *App) prepareListenData(addr string, isTls bool) ListenData {
```
Comment (cmd777) [inline listen.go:239]: Can you fix this linter warning?

File: listen.go [hunk 6]
```diff
@@ -231,8 +235,26 @@ func (app *App) ListenMutualTLSWithCertificate(addr string, cert tls.Certificate
...
+	return ListenData{
+		Host: host,
+		Port: port,
+		TLS:  isTls,
+	}
+}
+
 // startupMessage prepares the startup message with the handler number, port, address and other information
-func (app *App) startupMessage(addr string, tls bool, pids string) { //nolint: revive // Accepting a bool param is fine here
+func (app *App) startupMessage(addr string, isTls bool, pids string) { //nolint: revive // Accepting a bool param is fine here
```
Comment (cmd777) [inline listen.go:257]: And this as well

**Review by cmd777 — CHANGES_REQUESTED** [review]
Can you take a look at the new linter warnings/errors?

**Review by cmd777 — CHANGES_REQUESTED** [review]
Please revert the [f98133f](https://github.com/gofiber/fiber/pull/2491/commits/f98133f5755a8a6df70bf75e821a19b932669d33) commit.
While the go.mod does specify `go 1.20`, [tests still run with go 1.17-1.20](https://github.com/gofiber/fiber/blob/master/.github/workflows/test.yml#L20), and 1.17 does not contain `any` yet.

**Review by cmd777 — APPROVED** [review]

**Comment by efectn** [comment]
@Saman-Safaei can you add nolint flag for a method that isTLS param is used. 

You can check other nolint flag usages for an example

**Review by ReneWerner87 — APPROVED** [review]

**Review by efectn — APPROVED** [review]

**Pull request merged**